Add missing errors to our API docs and revamp the look of error tables for increased readability (#279)
alphagov/notifications-tech-docs: source/documentation/error_tables/_generic_errors.md

Do not use the content of the messages in your code. These can sometimes change, which may affect your API integration.

Use the status code or the error type instead, as these will not change.

### General errors

You may encounter following errors when making requests to a number of Notify's API endpoints.

Error message | How to fix
---|---
**BadRequestError (status code 400)**|

> `Can’t send to this recipient using a team-only API key.`|Use a live API key, or add recipient to `Guest list` (located in API Integration section)|

Reviewer: ```suggestion
`Cannot send to this recipient using a team-only API key.`|Use a live API key, or add recipient to `Guest list` (located in API Integration section)|
```

> `Can’t send to this recipient using a team-only API key.`|Use a live API key, or add recipient to `Guest list` (located in API Integration section)|

Reviewer: I thought I'd corrected it (though it might have been somewhere else!).

> `Can’t send to this recipient when service is in trial mode – see https://www.notifications.service.gov.uk/trial-mode`|You need to request for your service to go live before you can send messages to people outside your team.|

Reviewer: ```suggestion
`Cannot send to this recipient when service is in trial mode – see https://www.notifications.service.gov.uk/trial-mode`|You need to request for your service to go live before you can send messages to people outside your team.|
```

**BadRequestError (status code 403)**|
`Error: Your system clock must be accurate to within 30 seconds`|Check your system clock|
`Invalid token: API key not found`|Use the correct API key. Refer to [API keys](#api-keys) for more information|
**RateLimitError (status code 429)**|
`Exceeded rate limit for key type <team/test/live> of 3000/<custom limit> requests per 60 seconds`|Refer to [API rate limits](#rate-limits) for more information|
**TooManyRequestsError (status code 429)**|
`Exceeded send limits <your daily limit> for today`|Refer to [service limits](#daily-limits) for the limit number|
**Exception (status code 500)**|
`Internal server error`|Notify was unable to process the request, resend your notification.|

In addition to the above, you may also encounter:

> * various schema validation errors, for example when you forget to pass in an argument, or pass in an argument of a wrong type.

Reviewer 2: I wonder if it is worth giving a generic example of validation errors in keeping with the rest of the document, especially since examples are given for endpoint-specific errors.

Author: I see your point. At first I thought to include them, too. Chris pointed out to me that since these errors would be encountered in development stage of the integration (because they mean something is badly written), they would be handled manually, rather than programmatically. Hence, the errors that users will get should be clear enough on its own, and documenting them wouldn't bring much to the table. I mulled it over and ended up agreeing with him.

Author: Alright - let's do it then, but maybe as a separate PR, to get this one out of the door, ok? Since it's quite a big change - adding a sub-section for example validation errors, and linking it from every error table. I could make a card for it, so we don't forget.

Reviewer 3: Fine to do as a separate PR

* endpoint-specific errors, which are listed under each relevant API endpoint section.

Find references for endpoint-specific errors in:

> - [send a message](#send-a-message)

Reviewer 2: Nice idea attaching links to specific errors. Maybe the content could be updated to reflect the fact the links lead to the sections for the endpoints rather than directly to the error references.
I am mindful that for the links to go directly to the errors tat the error codes subsections would have to be given unique names.

Author: Interesting, yeah, I found this confusing, too, tbh.

Maybe we could change it to:

```suggestion
Find references for endpoint-specific errors at the end of:
- [send a message](#send-a-message)
```

@saimaghafoor any thoughts?

Author: For this one, I will change it for now to what I suggested, and when Saima is back, she could let us know what she thinks - it will be easy enough to change it back in case she is not happy with it.

Author: Actually, since each section, like "Send a message", has several error tables, my suggestion doesn't make sense. Let's leave it as is for now, and when Saima is back, maybe she will have some ideas that I could incorporate into the next PR?
